src/kernel/capinv.c

```diff
@@ -1409,6 +1422,21 @@ composite_syscall_slowpath(struct pt_regs *regs, int *thd_switch)
 			ret = tcap_delegate(tc, tcapsrc->tcap, res, prio);
 			if (unlikely(ret)) cos_throw(err, -EINVAL);
 
+			if(tcap_expended(tcap_current(cos_info))) {
+				struct thread *thd_next;
+				struct tcap *tc_curr, *tc_next;
+
+				tc_curr = tc_next = tcap_current(cos_info);
+				/* get the scheduler thread */
+				thd_next = thd_rcvcap_sched(tcap_rcvcap_thd(tc_curr));
+				assert(thd_next && thd_bound2rcvcap(thd_next) && thd_rcvcap_isreferenced(thd_next));
+
+				ret = cap_update(regs, thd, thd_next, tc_curr, tc_next, TCAP_TIME_NIL, ci, cos_info, 0);
+				if (unlikely(ret < 0)) cos_throw(err, ret);
+
+				*thd_switch = 1;
+			}
+
 			break;
 		}
 		case CAPTBL_OP_TCAP_DELEGATE:
```

Comment:
This has always been ridiculous to have in the switch.  Now it is quite complex so should obviously be a separate function.  You make the call when to do this.